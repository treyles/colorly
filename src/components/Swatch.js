import React from 'react';
import PropTypes from 'prop-types';
import { rgbToHex, parseRgb } from '../utils/helpers';

export default class Swatch extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      textColor: ''
    };

    this.handleClick = this.handleClick.bind(this);
    this.setInputRef = this.setInputRef.bind(this);
  }

  componentDidMount() {
    this.setTextContrast();
  }

  // check for sufficient contrast
  // between text and background color
  setTextContrast() {
    const rgb = parseRgb(this.props.color);
    const brightnessThreshold = 220;

    // algorithm to determine color brightness
    // https://www.w3.org/TR/AERT/#color-contrast
    const brightness = Math.round(
      (Number(rgb[0]) * 299 +
        Number(rgb[1]) * 587 +
        Number(rgb[2]) * 114) /
        1000
    );

    if (brightness > brightnessThreshold) {
      this.setState({ textColor: '#666' });
    }
  }

  setInputRef(ref) {
    this.hexInput = ref;
  }

  handleClick() {
    this.hexInput.select();
    document.execCommand('copy');
    this.props.activateCopyAlert();
  }

  render() {
    const hex = rgbToHex(this.props.color);

    return (
      <div
        className="swatch"
        style={{ background: `${hex}` }}
        onClick={this.handleClick}
        onKeyDown={this.handleClick}
        role="button"
        tabIndex={0}
      >
        <div className="pop-up" style={{ background: `${hex}` }} />
        <input
          ref={this.setInputRef}
          style={{ color: `${this.state.textColor}` }}
          value={`${hex}`}
          readOnly
        />
      </div>
    );
  }
}

Swatch.propTypes = {
  activateCopyAlert: PropTypes.func.isRequired,
  color: PropTypes.string.isRequired
};
