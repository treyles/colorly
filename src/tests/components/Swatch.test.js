import React from 'react';
import { shallow } from 'enzyme';
import Swatch from '../../components/Swatch';

describe('Swatch', () => {
  let component;
  const props = {
    activateCopyAlert: jest.fn(),
    color: ''
  };

  beforeEach(() => {
    component = shallow(<Swatch {...props} />);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('calls activateCopyAlert when swatch is clicked', () => {
    component.instance().hexInput = { select: jest.fn() };
    Object.defineProperty(window.document, 'execCommand', {
      value: jest.fn()
    });
    component.find('.swatch').simulate('click');
    expect(props.activateCopyAlert).toHaveBeenCalled();
  });

  it('sets textColor state to #666 when swatch color is light/white', () => {
    component.setProps({
      color: 'rgb(255, 255, 255)' // white
    });
    component.instance().setTextContrast();
    expect(component.state().textColor).toEqual('#666');
  });
});
