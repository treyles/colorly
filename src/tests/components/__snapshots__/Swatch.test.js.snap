// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swatch renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Swatch
    activateCopyAlert={[MockFunction]}
    color=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="pop-up"
          style={
            Object {
              "background": "#NAN",
            }
          }
        />,
        <input
          readOnly={true}
          style={
            Object {
              "color": "",
            }
          }
          value="#NAN"
        />,
      ],
      "className": "swatch",
      "onClick": [Function],
      "onKeyDown": [Function],
      "role": "button",
      "style": Object {
        "background": "#NAN",
      },
      "tabIndex": 0,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "pop-up",
          "style": Object {
            "background": "#NAN",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "readOnly": true,
          "style": Object {
            "color": "",
          },
          "value": "#NAN",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="pop-up"
            style={
              Object {
                "background": "#NAN",
              }
            }
          />,
          <input
            readOnly={true}
            style={
              Object {
                "color": "",
              }
            }
            value="#NAN"
          />,
        ],
        "className": "swatch",
        "onClick": [Function],
        "onKeyDown": [Function],
        "role": "button",
        "style": Object {
          "background": "#NAN",
        },
        "tabIndex": 0,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "pop-up",
            "style": Object {
              "background": "#NAN",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "readOnly": true,
            "style": Object {
              "color": "",
            },
            "value": "#NAN",
          },
          "ref": [Function],
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
